taichi/system/timeline.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <vector>
+#include <mutex>
+
+#include "taichi/common/core.h"
+#include "taichi/system/timer.h"
+
+TI_NAMESPACE_BEGIN
+
+struct TimelineEvent {
+  std::string name;
+  bool begin;
+  float64 time;
+  std::string tid;
+
+  std::string to_json();
+};
+
+class Timeline {
+ public:
+  Timeline();
+
+  ~Timeline();
+
+  static Timeline &get_this_thread_instance();
+
+  void set_name(const std::string &tid) {
+    tid_ = tid;
+  }
+
+  std::string get_name() {
+    return tid_;
+  }
+
+  void clear();
+
+  void insert_event(const TimelineEvent &e);
+
+  std::vector<TimelineEvent> fetch_events();
+
+  class Guard {
+   public:
+    Guard(const std::string &name);
+
+    ~Guard();
+
+   private:
+    std::string name_;
+  };
+
+ private:
+  std::string tid_;
+  std::mutex mut_;
+  std::vector<TimelineEvent> events_;
+};
+
+// A timeline system for multi-threaded applications
+class Timelines {
+ public:
+  static Timelines &get_instance();
+
+  void insert_events(const std::vector<TimelineEvent> &events,
+                     bool lock = true);
+
+  void insert_timeline(Timeline *timeline);
+
+  void remove_timeline(Timeline *timeline);
+
+  void clear();
+
+  void save(const std::string &filename);
+
+  bool get_enabled();
+
+  void set_enabled(bool enabled);
+
+ private:
+  std::mutex mut_;
+  std::vector<TimelineEvent> events_;
+  std::vector<Timeline *> timelines_;
+  bool enabled_{false};
+};
+
+#define TI_TIMELINE(name) \
+  taichi::Timeline::Guard _timeline_guard_##__LINE__(name);
+
+#define TI_AUTO_TIMELINE TI_TIMELINE(__FUNCTION__)
```

Comment:
Let's hope this won't introduce the overhead we've seen in `TI_AUTO_PROF`...

Reply:
So far if the user does not explicitly turn on `ti.init(timeline=True)` the overhead would be negligiable:

https://github.com/taichi-dev/taichi/pull/2164/files#diff-b049e9e32375858a4d5ced885fc11e42177feb5384bf74e8c900435a31755a05R36